### PR TITLE
sunrise-router

### PR DESCRIPTION
**Pull request by dominikzogg**

### CODE REVIEW

**Comment by WinterSilence** [comment]
bad solution and forget "final" for classes

File: src/Router/SunriseRouter.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        try {
+            $sunriseRoute = $this->router->match($request);
+
+            /** @var RouteInterface $route */
+            $route = $this->routes[$sunriseRoute->getName()];
+
+            return $route->withAttributes($sunriseRoute->getAttributes());
+        } catch (RouteNotFoundException $exception) {
+            throw RouterException::createForNotFound($request->getRequestTarget());
```
Comment (WinterSilence) [inline src/Router/SunriseRouter.php]: separate exception and his factory, see PSR-17 for example

File: src/Router/SunriseRouter.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        }
+    }
+
+    /**
+     * @param array<string, string> $attributes
+     * @param array<string, mixed>  $queryParams
+     *
+     * @throws RouterException
+     */
+    public function generateUrl(
```
Comment (WinterSilence) [inline src/Router/SunriseRouter.php]: If u use only UriInterface instance, then why sends RequestInstance? Router can work with external requests too.

File: src/Router/SunriseRouter.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        return $uri->getScheme().'://'.$uri->getAuthority().$requestTarget;
+    }
+
+    /**
+     * @param array<string, string> $attributes
+     * @param array<string, mixed>  $queryParams
+     *
+     * @throws RouterException
+     */
+    public function generatePath(string $name, array $attributes = [], array $queryParams = []): string
```
Comment (WinterSilence) [inline src/Router/SunriseRouter.php]: query not path part

File: src/Router/SunriseRouter.php [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        return '/'; // todo implement
+    }
+
+    /**
+     * @param array<RouteInterface> $routes
+     *
+     * @return array<RouteInterface>
+     */
+    private function getRoutesByName(array $routes): array
```
Comment (WinterSilence) [inline src/Router/SunriseRouter.php]: method unuseful, u can strore routes with name keys

File: src/Router/SunriseRouter.php [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    private function createRouter(array $routes): Router
+    {
+        $router = new Router();
+
+        foreach ($routes as $route) {
+            $router->route(
+                $route->getName(),
+                $route->getPath(),
+                [$route->getMethod()],
+                new class() implements RequestHandlerInterface {
```
Comment (WinterSilence) [inline src/Router/SunriseRouter.php]: too strange solution, RequestHandlerInterface must be passed in constructor

**Comment by dominikzogg** [comment]
@WinterSilence 

First of all, try to understand the use case before criticize a solution. its not the goal to use this router exclusive, it only an adapter besides others. The router (how its build can be used as a framework replacement, but that's not the goal here.

About final classes, there is no reason for not having final as a default and there are good talks out there, which will explain you why.

**Comment by WinterSilence** [comment]
> About final classes, there is no reason for not having final as a default and there are good talks out there, which will explain you why.

proof link pls

**Comment by dominikzogg** [comment]
> proof link pls

https://matthiasnoback.nl/2018/09/final-classes-by-default-why/
https://ocramius.github.io/blog/when-to-declare-classes-final/

related:

https://en.wikipedia.org/wiki/Composition_over_inheritance
https://javarevisited.blogspot.com/2013/06/why-favor-composition-over-inheritance-java-oops-design.html

**Comment by WinterSilence** [comment]
@dominikzogg Your class does not meet the conditions in this post.

**Comment by fenric** [comment]
@dominikzogg сan I ask a question? please tell me... what do you think about `RouteCollection`? and generally about the router? I would be very grateful for your assessment...

**Comment by dominikzogg** [comment]
@fenric in my point of view v2 of the router is ready

**Pull request merged**